### ISSUE TITLE
GHC environment files can break GHCi

### ISSUE THREAD

**Issue by snoyberg** [haskell/cabal#5988]
Moving from commercialhaskell/stack#4706.

Following the same steps for repro in [this comment](https://github.com/commercialhaskell/stack/issues/4706#issuecomment-480584940), but leaving out Stack entirely, results in `ghci` invocation failing. Demonstration:

```
+ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.6.3
+ cabal --version
cabal-install version 2.4.1.0
compiled using version 2.4.1.0 of the Cabal library 
+ echo 'cabal-version: 2.4'
+ echo 'name: example'
+ echo 'version: 0'
+ echo 'library { default-language: Haskell2010 }'
+ cabal v2-build
Resolving dependencies...
Build profile: -w ghc-8.6.3 -O1
In order, the following will be built (use -v for more details):
 - example-0 (lib) (first run)
Configuring library for example-0..
Preprocessing library for example-0..
Building library for example-0..
+ ghci
GHCi, version 8.6.3: http://www.haskell.org/ghc/  :? for help
Loaded package environment from /Users/michael/Desktop/4706-example/tmp/.ghc.environment.x86_64-darwin-8.6.3

<interactive>:1:6: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.

<interactive>:1:30: error:
    Not in scope: ‘System.IO.stdin’
    No module named ‘System.IO’ is imported.

<interactive>:1:46: error:
    Not in scope: data constructor ‘System.IO.NoBuffering’
    No module named ‘System.IO’ is imported.

<interactive>:1:70: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.

<interactive>:1:94: error:
    Not in scope: ‘System.IO.stdout’
    No module named ‘System.IO’ is imported.

<interactive>:1:111: error:
    Not in scope: data constructor ‘System.IO.NoBuffering’
    No module named ‘System.IO’ is imported.

<interactive>:1:135: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.

<interactive>:1:159: error:
    Not in scope: ‘System.IO.stderr’
    No module named ‘System.IO’ is imported.

<interactive>:1:176: error:
    Not in scope: data constructor ‘System.IO.NoBuffering’
    No module named ‘System.IO’ is imported.
```

The problem is that the environment file is created without `base` in it. However, `ghci` itself relies on functionality present in `base`.

Alternative to a Cabal bug, it could be argued that:

* This is a GHC bug, since `ghci` should be able to load up the functionality it needs from `base` regardless of package settings
* This isn't a bug at all; having a package without a dependency on `base` is a corner case that isn't addressed by the environment file writing code

**Comment by phadej** [comment]
With `composition` package (which doesn't depend on anything, not even `base`) both `cabal` and `stack` fail:

```
composition-1.0.2.1 % cabal v2-repl
Resolving dependencies...
Build profile: -w ghc-8.4.4 -O1
In order, the following will be built (use -v for more details):
 - composition-1.0.2.1 (lib) (first run)
Preprocessing library for composition-1.0.2.1..
GHCi, version 8.4.4: http://www.haskell.org/ghc/  :? for help

<interactive>:1:6: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.
```

```
composition-1.0.2.1 % cat stack.yaml 
resolver: lts-12.26
packages: [.]

composition-1.0.2.1 % stack repl    
composition-1.0.2.1: initial-build-steps (lib)
Configuring GHCi with the following packages: composition
GHCi, version 8.4.4: http://www.haskell.org/ghc/  :? for help

<interactive>:1:6: error:
    Not in scope: ‘System.IO.hSetBuffering’
    No module named ‘System.IO’ is imported.
```

---

So I argue it's it's not a bug in `cabal`. Especially as environment does contain global package database:

```
clear-package-db
global-package-db
package-db /home/phadej/.cabal/store/ghc-8.4.4/package.db
package-db dist-newstyle/packagedb/ghc-8.4.4
```

(and it does for `stack` too). So, as you said, either `ghci` should be able to load needed stuff from `base` in the package dbs passed to it, even `base` is hidden; or then this is just a corner case.

---

Note: maybe something is changed in GHC, as ghc-7.10.3 works (base is hidden, but stuff loads).
Also I don't think that change was related to environment files, as some environment file functionality is already in GHC-7.10 IIRC (just not good enough to be used). 

```
composition-1.0.2.1% cabal new-repl -w ghc-7.10.3
Resolving dependencies...
Build profile: -w ghc-7.10.3 -O1
In order, the following will be built (use -v for more details):
 - composition-1.0.2.1 (lib) (first run)
Preprocessing library for composition-1.0.2.1..
GHCi, version 7.10.3: http://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling Data.Composition ( src/Data/Composition.hs, interpreted )
Ok, modules loaded: Data.Composition.
*Data.Composition> :m +Data.List

<no location info>:
    Could not find module ‘Data.List’
    It is a member of the hidden package ‘base-4.8.2.0’.
    Perhaps you need to add ‘base’ to the build-depends in your .cabal file.
```

**Comment by 23Skidoo** [comment]
We should file a ticket at the GHC bug tracker, but if anyone is willing to write a patch that makes env files always include `base` (maybe only for the affected GHC versions), I think it'd be accepted.

**Comment by snoyberg** [comment]
> a patch that makes env files always include base

I'm not an expert on the features involved, but I think this may prove to cause breakage. One possibility would be if another package exports the same modules as `base`. That may sound strange at first, but it's exactly be [`rerebase`](https://www.stackage.org/package/rerebase) does, for instance.

I'd be happy to open up an issue with GHC, I'll comment again here after that's done.

**Comment by snoyberg** [comment]
Issue opened upstream: https://gitlab.haskell.org/ghc/ghc/issues/16563

**Comment by 23Skidoo** [comment]
> a patch that makes env files always include base

Upd.: @hvr also thinks it's a bad idea, but he can't comment In this thread because he's blocked by @snoyberg (and vice versa).

**Comment by 23Skidoo** [comment]
> rerebase

Another quite common case is having a module called `Prelude`, see `base-noprelude`.

**Comment by 23Skidoo** [comment]
I'm closing this ticket in favour of the upstream one, since we all seem to agree that always adding `base` to the env file is a bad idea and I can't think of any other workaround. If anyone disagrees, please reopen or open a separate enhancement request.

**Comment by snoyberg** [comment]
SGTM, thanks @23skidoo

**Comment by ulysses4ever** [comment]
I have Cabal 2.4.1 and GHC 8.6.4, and this thing bites me when trying `cabal new-repl` even on a toy project. Is there any workaround I can apply to start the repl?
